File: scripts/fetch_benchmark_datasets.py
# /// script
# dependencies = [
#     "pyarrow",
#     "requests",
# ]
# ///

"""
Fetch and verify fixed public benchmark datasets from Hugging Face Hub.
Tracks provenance including exact commit SHAs, URLs, licenses, SHA-256 digests,
and Parquet schema profiles for Git LFS.
"""

import hashlib
import json
import os
import sys
from datetime import datetime, timezone
import pyarrow.parquet as pq
import requests

BENCHMARK_DATA_DIR = os.path.abspath(os.path.join(os.path.dirname(__file__), "..", "benchmarks", "data"))

DATASET_CONFIGS = [
    {
        "id": "tpch-lineitem-sf001",
        "filename": "tpch_lineitem_sf001.parquet",
        "description": "TPC-H Benchmark lineitem table at Scale Factor 0.01 (standard analytical decision support benchmark).",
        "source_platform": "Hugging Face Hub",
        "upstream_repository": "liangyc/tpch-sf-0_01",
        "upstream_commit": "a91e9442ea9dc1fe4729db3138afdd03ef36a9e5",
        "source_url": "https://huggingface.co/datasets/liangyc/tpch-sf-0_01/resolve/a91e9442ea9dc1fe4729db3138afdd03ef36a9e5/lineitem/train/0000.parquet",
        "license": "Apache-2.0"
    },
    {
        "id": "adult-census-income",
        "filename": "adult_census_income.parquet",
        "description": "Adult Census Income dataset from UCI Machine Learning Repository (standard categorical tabular benchmark).",
        "source_platform": "Hugging Face Hub",
        "upstream_repository": "scikit-learn/adult-census-income",
        "upstream_commit": "aefa0f0f1b03a11dd48f460913e20a1d50b4e53c",
        "source_url": "https://huggingface.co/datasets/scikit-learn/adult-census-income/resolve/aefa0f0f1b03a11dd48f460913e20a1d50b4e53c/default/train/0000.parquet",
        "license": "CC-BY-4.0"
    },
    {
        "id": "diamonds",
        "filename": "diamonds.parquet",
        "description": "Diamonds dataset from ggplot2 (standard regression tabular benchmark with numeric & ordinal features).",
        "source_platform": "Hugging Face Hub",
        "upstream_repository": "inria-soda/tabular-benchmark",
        "upstream_commit": "cb2bcee34f8fbd271c04dc644fd91f23a51a8570",
        "source_url": "https://huggingface.co/datasets/inria-soda/tabular-benchmark/resolve/cb2bcee34f8fbd271c04dc644fd91f23a51a8570/reg_cat_diamonds/train/0000.parquet",
        "license": "CC0-1.0"
    }
]


def download_file(url: str, dest_path: str) -> str:
    print(f"📥 Fetching: {url} -> {dest_path}")
    headers = {"User-Agent": "Parquet.SourceGenerator-Benchmark-Sync/1.0"}
    resp = requests.get(url, headers=headers, stream=True, timeout=120)
    resp.raise_for_status()

    hasher = hashlib.sha256()
    total_bytes = 0
    with open(dest_path, "wb") as f:
        for chunk in resp.iter_content(chunk_size=65536):
            if chunk:
                f.write(chunk)
                hasher.update(chunk)
                total_bytes += len(chunk)

    digest = hasher.hexdigest()
    print(f"   Done ({total_bytes:,} bytes, SHA-256: {digest})")
    return digest


def inspect_parquet_file(file_path: str):
    parquet_file = pq.ParquetFile(file_path)
    metadata = parquet_file.metadata
    schema = parquet_file.schema

    num_rows = metadata.num_rows
    num_columns = metadata.num_columns
    num_row_groups = metadata.num_row_groups
    format_version = metadata.format_version

    columns = []
    dict_columns = []
    plain_columns = []

    # Inspect first row group columns
    rg = metadata.row_group(0)
    for c_idx in range(rg.num_columns):
        col_meta = rg.column(c_idx)
        col_name = col_meta.path_in_schema
        physical_type = col_meta.physical_type
        encodings = [str(enc) for enc in col_meta.encodings]
        is_dict = any("RLE_DICTIONARY" in enc or "PLAIN_DICTIONARY" in enc for enc in encodings)
        if is_dict:
            dict_columns.append(col_name)
        else:
            plain_columns.append(col_name)

        col_schema = schema.column(c_idx)
        columns.append({
            "name": col_name,
            "physical_type": str(physical_type),
            "logical_type": str(col_schema.logical_type) if col_schema.logical_type else None,
            "encodings": encodings,
            "dictionary_encoded": is_dict
        })

    compression = rg.column(0).compression if rg.num_columns > 0 else "UNKNOWN"

    return {
        "row_count": num_rows,
        "column_count": num_columns,
        "row_group_count": num_row_groups,
        "format_version": format_version,
        "compression": compression.lower(),
        "columns": columns,
        "dictionary_encoded_columns": dict_columns,
        "plain_columns": plain_columns
    }


def main():
    os.makedirs(BENCHMARK_DATA_DIR, exist_ok=True)
    manifest = {
        "$schema": "https://json-schema.org/draft/2020-12/schema",
        "description": "Cryptographic Data Provenance Manifest for Parquet.SourceGenerator Benchmarks",
        "generated_at": datetime.now(timezone.utc).isoformat(),
        "datasets": []
    }

    for cfg in DATASET_CONFIGS:
        dest_path = os.path.join(BENCHMARK_DATA_DIR, cfg["filename"])
        
        # Download file
        sha256 = download_file(cfg["source_url"], dest_path)
        file_size = os.path.getsize(dest_path)

        # Inspect parquet profile
        profile = inspect_parquet_file(dest_path)

        entry = {
            "id": cfg["id"],
            "filename": cfg["filename"],
            "description": cfg["description"],
            "provenance": {
                "source_platform": cfg["source_platform"],
                "upstream_repository": cfg["upstream_repository"],
                "upstream_commit": cfg["upstream_commit"],
                "source_url": cfg["source_url"],
                "license": cfg["license"],
                "verified_at": datetime.now(timezone.utc).isoformat()
            },
            "integrity": {
                "file_size_bytes": file_size,
                "sha256": sha256,
                "git_lfs_oid": f"sha256:{sha256}"
            },
            "parquet_profile": profile
        }
        manifest["datasets"].append(entry)

    manifest_path = os.path.join(BENCHMARK_DATA_DIR, "provenance.json")
    with open(manifest_path, "w", encoding="utf-8") as f:
        json.dump(manifest, f, indent=2)
    print(f"\n📄 Saved provenance manifest to {manifest_path}")

    # Generate Markdown summary table
    md_path = os.path.join(BENCHMARK_DATA_DIR, "PROVENANCE.md")
    with open(md_path, "w", encoding="utf-8") as f:
        f.write("# Benchmark Dataset Provenance & Integrity Register\n\n")
        f.write("This directory contains source-controlled Parquet datasets managed via **Git LFS** for reliable, deterministic, and offline-capable continuous integration (CI) and performance benchmarking.\n\n")
        f.write("## 📦 Registered Datasets\n\n")
        f.write("| Dataset | File | Rows | Size | Codec | Dict Columns | Upstream Repo | License | SHA-256 |\n")
        f.write("|:---|:---|:---:|:---:|:---:|:---:|:---|:---:|:---:|\n")
        for ds in manifest["datasets"]:
            prof = ds["parquet_profile"]
            integ = ds["integrity"]
            prov = ds["provenance"]
            size_kb = integ["file_size_bytes"] / 1024.0
            dict_count = len(prof["dictionary_encoded_columns"])
            total_cols = prof["column_count"]
            short_sha = integ["sha256"][:12] + "..."
            f.write(f"| **{ds['id']}** | `{ds['filename']}` | {prof['row_count']:,} | {size_kb:.1f} KB | `{prof['compression']}` | {dict_count}/{total_cols} | [{prov['upstream_repository']}](https://huggingface.co/datasets/{prov['upstream_repository']}) | `{prov['license']}` | `{short_sha}` |\n")
        
        f.write("\n## 🔍 Schema & Encoding Breakdown\n\n")
        for ds in manifest["datasets"]:
            prof = ds["parquet_profile"]
            integ = ds["integrity"]
            prov = ds["provenance"]
            f.write(f"### {ds['id']} (`{ds['filename']}`)\n\n")
            f.write(f"- **Description**: {ds['description']}\n")
            f.write(f"- **Upstream Source**: [{prov['upstream_repository']} (commit `{prov['upstream_commit'][:10]}`)](https://huggingface.co/datasets/{prov['upstream_repository']})\n")
            f.write(f"- **Download URL**: `{prov['source_url']}`\n")
            f.write(f"- **License**: `{prov['license']}`\n")
            f.write(f"- **File Size**: {integ['file_size_bytes']:,} bytes\n")
            f.write(f"- **SHA-256**: `{integ['sha256']}`\n")
            f.write(f"- **Row Count**: {prof['row_count']:,}\n")
            f.write(f"- **Row Groups**: {prof['row_group_count']}\n")
            f.write(f"- **Compression**: `{prof['compression']}`\n\n")
            f.write("| Column Name | Physical Type | Logical Type | Encoded via Dictionary? |\n")
            f.write("|:---|:---|:---|:---:|\n")
            for col in prof["columns"]:
                is_dict = "✅ Yes" if col["dictionary_encoded"] else "❌ Plain"
                logical = f"`{col['logical_type']}`" if col["logical_type"] else "—"
                f.write(f"| `{col['name']}` | `{col['physical_type']}` | {logical} | {is_dict} |\n")
            f.write("\n")

    print(f"📝 Saved provenance markdown register to {md_path}")


if __name__ == "__main__":
    main()

File: scripts/FetchBenchmarkDatasets.cs
#:package Parquet.Net@6.1.0

using System;
using System.Collections.Generic;
using System.IO;
using System.Linq;
using System.Net.Http;
using System.Security.Cryptography;
using System.Text.Json;
using System.Threading.Tasks;
using Parquet;
using Parquet.SourceGenerator.Scripts;

// -----------------------------------------------------------------------------
// FetchBenchmarkDatasets.cs
// Standalone C# script (.NET 10) to fetch and verify fixed public benchmark
// datasets from Hugging Face Hub.
// Tracks cryptographic provenance (SHA-256, commit SHAs, schema profiles).
// -----------------------------------------------------------------------------

var configs = new List<DatasetConfig>
{
    new(
        Id: "tpch-lineitem-sf001",
        Filename: "tpch_lineitem_sf001.parquet",
        Description: "TPC-H Benchmark lineitem table at Scale Factor 0.01 (standard analytical decision support benchmark).",
        SourcePlatform: "Hugging Face Hub",
        UpstreamRepository: "liangyc/tpch-sf-0_01",
        UpstreamCommit: "a91e9442ea9dc1fe4729db3138afdd03ef36a9e5",
        SourceUrl: "https://huggingface.co/datasets/liangyc/tpch-sf-0_01/resolve/a91e9442ea9dc1fe4729db3138afdd03ef36a9e5/lineitem/train/0000.parquet",
        License: "Apache-2.0"
    ),
    new(
        Id: "adult-census-income",
        Filename: "adult_census_income.parquet",
        Description: "Adult Census Income dataset from UCI Machine Learning Repository (standard categorical tabular benchmark).",
        SourcePlatform: "Hugging Face Hub",
        UpstreamRepository: "scikit-learn/adult-census-income",
        UpstreamCommit: "aefa0f0f1b03a11dd48f460913e20a1d50b4e53c",
        SourceUrl: "https://huggingface.co/datasets/scikit-learn/adult-census-income/resolve/aefa0f0f1b03a11dd48f460913e20a1d50b4e53c/default/train/0000.parquet",
        License: "CC-BY-4.0"
    ),
    new(
        Id: "diamonds",
        Filename: "diamonds.parquet",
        Description: "Diamonds dataset from ggplot2 (standard regression tabular benchmark with numeric & ordinal features).",
        SourcePlatform: "Hugging Face Hub",
        UpstreamRepository: "inria-soda/tabular-benchmark",
        UpstreamCommit: "cb2bcee34f8fbd271c04dc644fd91f23a51a8570",
        SourceUrl: "https://huggingface.co/datasets/inria-soda/tabular-benchmark/resolve/cb2bcee34f8fbd271c04dc644fd91f23a51a8570/reg_cat_diamonds/train/0000.parquet",
        License: "CC0-1.0"
    ),
};

bool force = args.Contains("--force", StringComparer.OrdinalIgnoreCase);
string dataDir = FindBenchmarkDataDir();
Directory.CreateDirectory(dataDir);

Console.WriteLine($"📦 Managing benchmark datasets in: {dataDir}");

using var httpClient = new HttpClient();
httpClient.DefaultRequestHeaders.UserAgent.ParseAdd("Parquet.SourceGenerator-Benchmark-Sync/1.0");

var datasetSummaries = new List<DatasetSummary>();

foreach (var cfg in configs)
{
    string targetPath = Path.Combine(dataDir, cfg.Filename);
    bool shouldDownload =
        force || !System.IO.File.Exists(targetPath) || new FileInfo(targetPath).Length < 1024;

    if (shouldDownload)
    {
        Console.WriteLine($"📥 Downloading {cfg.Id} from {cfg.SourceUrl}...");
        using var response = await httpClient.GetAsync(
            cfg.SourceUrl,
            HttpCompletionOption.ResponseHeadersRead
        );
        response.EnsureSuccessStatusCode();

        await using var sourceStream = await response.Content.ReadAsStreamAsync();
        await using var fileStream = System.IO.File.Create(targetPath);
        await sourceStream.CopyToAsync(fileStream);
        Console.WriteLine($"   Saved {cfg.Filename} ({new FileInfo(targetPath).Length:N0} bytes)");
    }
    else
    {
        Console.WriteLine(
            $"✅ {cfg.Filename} already present ({new FileInfo(targetPath).Length:N0} bytes)"
        );
    }

    // Hash file
    string sha256;
    long fileSizeBytes;
    using (var fs = System.IO.File.OpenRead(targetPath))
    {
        fileSizeBytes = fs.Length;
        byte[] hash = SHA256.HashData(fs);
        sha256 = Convert.ToHexString(hash).ToLowerInvariant();
    }

    // Inspect Parquet schema and metadata
    long rowCount;
    int columnCount;
    int rowGroupCount;
    string compression = "none";
    var columns = new List<ColumnSummary>();
    var dictColumns = new List<string>();
    var plainColumns = new List<string>();

    using (var fs = System.IO.File.OpenRead(targetPath))
    {
        await using var reader = await ParquetReader.CreateAsync(fs);
        rowCount = reader.Metadata?.NumRows ?? 0;
        rowGroupCount = reader.RowGroupCount;
        columnCount = reader.Schema.Fields.Count;

        var firstRg = reader.Metadata?.RowGroups.FirstOrDefault();
        if (firstRg != null)
        {
            compression =
                firstRg.Columns.FirstOrDefault()?.MetaData?.Codec.ToString().ToLowerInvariant()
                ?? "none";

            for (int i = 0; i < firstRg.Columns.Count; i++)
            {
                var col = firstRg.Columns[i];
                var meta = col.MetaData;
                string colName =
                    meta != null
                        ? string.Join(".", meta.PathInSchema)
                        : reader.Schema.Fields[i].Name;
                string physicalType = meta?.Type.ToString() ?? "UNKNOWN";
                var encodings = meta?.Encodings.Select(e => e.ToString()).ToList() ?? [];
                bool isDict = encodings.Any(e =>
                    e.Contains("DICTIONARY", StringComparison.OrdinalIgnoreCase)
                );

                if (isDict)
                    dictColumns.Add(colName);
                else
                    plainColumns.Add(colName);

                columns.Add(new ColumnSummary(colName, physicalType, encodings, isDict));
            }
        }
    }

    datasetSummaries.Add(
        new DatasetSummary(
            Config: cfg,
            FileSizeBytes: fileSizeBytes,
            Sha256: sha256,
            RowCount: rowCount,
            ColumnCount: columnCount,
            RowGroupCount: rowGroupCount,
            Compression: compression,
            DictColumns: dictColumns,
            PlainColumns: plainColumns,
            Columns: columns
        )
    );

    Console.WriteLine($"   SHA-256: {sha256}");
    Console.WriteLine(
        $"   Rows: {rowCount:N0}, Cols: {columnCount}, Compression: {compression}, Dictionaries: {dictColumns.Count}"
    );
}

string manifestPath = Path.Combine(dataDir, "provenance.json");
using (var fs = System.IO.File.Create(manifestPath))
using (var writer = new Utf8JsonWriter(fs, new JsonWriterOptions { Indented = true }))
{
    writer.WriteStartObject();
    writer.WriteString("$schema", "https://json-schema.org/draft/2020-12/schema");
    writer.WriteString(
        "description",
        "Cryptographic Data Provenance Manifest for Parquet.SourceGenerator Benchmarks"
    );
    writer.WriteString("generated_at", DateTime.UtcNow.ToString("o"));

    writer.WriteStartArray("datasets");
    foreach (var ds in datasetSummaries)
    {
        writer.WriteStartObject();
        writer.WriteString("id", ds.Config.Id);
        writer.WriteString("filename", ds.Config.Filename);
        writer.WriteString("description", ds.Config.Description);

        writer.WriteStartObject("provenance");
        writer.WriteString("source_platform", ds.Config.SourcePlatform);
        writer.WriteString("upstream_repository", ds.Config.UpstreamRepository);
        writer.WriteString("upstream_commit", ds.Config.UpstreamCommit);
        writer.WriteString("source_url", ds.Config.SourceUrl);
        writer.WriteString("license", ds.Config.License);
        writer.WriteString("verified_at", DateTime.UtcNow.ToString("o"));
        writer.WriteEndObject();

        writer.WriteStartObject("integrity");
        writer.WriteNumber("file_size_bytes", ds.FileSizeBytes);
        writer.WriteString("sha256", ds.Sha256);
        writer.WriteString("git_lfs_oid", $"sha256:{ds.Sha256}");
        writer.WriteEndObject();

        writer.WriteStartObject("parquet_profile");
        writer.WriteNumber("row_count", ds.RowCount);
        writer.WriteNumber("column_count", ds.ColumnCount);
        writer.WriteNumber("row_group_count", ds.RowGroupCount);
        writer.WriteString("compression", ds.Compression);

        writer.WriteStartArray("dictionary_columns");
        foreach (var c in ds.DictColumns)
            writer.WriteStringValue(c);
        writer.WriteEndArray();

        writer.WriteStartArray("plain_columns");
        foreach (var c in ds.PlainColumns)
            writer.WriteStringValue(c);
        writer.WriteEndArray();

        writer.WriteStartArray("columns");
        foreach (var c in ds.Columns)
        {
            writer.WriteStartObject();
            writer.WriteString("name", c.Name);
            writer.WriteString("physical_type", c.PhysicalType);

            writer.WriteStartArray("encodings");
            foreach (var enc in c.Encodings)
                writer.WriteStringValue(enc);
            writer.WriteEndArray();

            writer.WriteBoolean("dictionary_encoded", c.DictionaryEncoded);
            writer.WriteEndObject();
        }
        writer.WriteEndArray();

        writer.WriteEndObject(); // parquet_profile
        writer.WriteEndObject(); // dataset
    }
    writer.WriteEndArray(); // datasets
    writer.WriteEndObject(); // root
}

Console.WriteLine($"\n📝 Updated provenance manifest: {manifestPath}");
return 0;

static string FindBenchmarkDataDir()
{
    string[] candidates =
    [
        Path.Combine(Directory.GetCurrentDirectory(), "benchmarks", "data"),
        Path.Combine(AppContext.BaseDirectory, "..", "..", "..", "..", "benchmarks", "data"),
        Path.Combine(AppContext.BaseDirectory, "benchmarks", "data"),
        Path.Combine(AppContext.BaseDirectory, "data"),
    ];

    foreach (var candidate in candidates)
    {
        string full = Path.GetFullPath(candidate);
        if (Directory.Exists(full))
        {
            return full;
        }
    }

    return Path.GetFullPath(Path.Combine("benchmarks", "data"));
}

namespace Parquet.SourceGenerator.Scripts
{
    internal sealed record DatasetConfig(
        string Id,
        string Filename,
        string Description,
        string SourcePlatform,
        string UpstreamRepository,
        string UpstreamCommit,
        string SourceUrl,
        string License
    );

    internal sealed record ColumnSummary(
        string Name,
        string PhysicalType,
        IReadOnlyList<string> Encodings,
        bool DictionaryEncoded
    );

    internal sealed record DatasetSummary(
        DatasetConfig Config,
        long FileSizeBytes,
        string Sha256,
        long RowCount,
        int ColumnCount,
        int RowGroupCount,
        string Compression,
        IReadOnlyList<string> DictColumns,
        IReadOnlyList<string> PlainColumns,
        IReadOnlyList<ColumnSummary> Columns
    );
}
